earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp

```diff
@@ -278,16 +291,76 @@ class StateUpdater::SetStateVisitor : public default_dfs_visitor {
     }
 };
 
-void StateUpdater::SetStateForRefAndDependents(
+
+class StateUpdater::SetBlockingStateVisitor : public StateUpdater::VisitorBase {
+  private:
+    using BlockedSet = std::unordered_set<SharedString>;
+
+    // This is a shared_ptr because the visitor will be copied several times
+    // during the course of the traversal.
+    const std::shared_ptr<BlockedSet> hasBlockingInputs;
+    const std::shared_ptr<BlockedSet> hasBlockingChildren;
+
+    void MarkParentsForLaterProcessing(AssetHandle<AssetVersionImpl> & version) const override {
+      if (AssetDefs::IsBlocking(newState)){
```

Comment:
Minor: hopefully you wouldn't use this class at all unless newState is Blocking. Maybe make this an assert in the constructor?